NS for FX: add test for a simple sparsenn model

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52402 ns for fx: update graph matching to not match nodes with equal types
* #52395 ns for fx: support linear_relu for weight matching
* #52368 ns for fx: allow graph matching of parents of cat
* #52357 ns for fx: make unshadowed activation comparison work for N models
* #52356 ns for fx: make weights comparison work on N models
* #52130 ns for fx: add support for subgraph matching
* **#52092 NS for FX: add test for a simple sparsenn model**

Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403095](https://our.internmc.facebook.com/intern/diff/D26403095)

vkuzo committed:
NS for FX: add test for a simple sparsenn model

Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403095](https://our.internmc.facebook.com/intern/diff/D26403095)

[ghstack-poisoned]

vkuzo committed:
Update on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403095](https://our.internmc.facebook.com/intern/diff/D26403095)

[ghstack-poisoned]

raghuramank100 commented on test/quantization/test_numeric_suite_fx.py:
```diff
@@ -771,3 +771,151 @@ def forward(self, x):
+    """
+
+    def _get_sparsenn_toy_model(self):
+        class DenseTopMLP(nn.Module):
```
NIT: Can we move all models to https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py#L865

raghuramank100 commented on torch/quantization/ns/graph_passes.py:
```diff
@@ -120,7 +121,7 @@ def _insert_dtype_cast_after_node(
     new_dtype_cast_name = \
         get_new_attr_name_with_prefix(node_name_prefix)(gm_b)
-    return prev_node_c.graph.create_node(
+    return graph_c.create_node(
```
Do you need graph_c to be different from the one containing prev_node_c?

raghuramank100 commented on torch/quantization/ns/utils.py:
```diff
@@ -34,14 +35,17 @@ class NodeIOType(enum.Enum):
         fp32_fun_target_names = ('torch.nn.functional', 'torch.nn')
+        # hack alert: this is not ready for production
+        # TODO(future PR): use a real mapping
+        fp32_funs = (torch.cat,)
```
Doesnt torch.cat support both fp32 and quantized ops? My understanding is that the dispatcher handles dtype specific calls for the same function name.  For figuring out the i.o type, do you need to look at the dtypes of the inputs and outputs for a node?

vkuzo replied:
Yeah, the inside of this function is not ready for a wide variety for users, for sure.  There other hacks in here which are just as bad / worse :). I'm punting a better implementation of the insides of this function until the future.

We should align on the right interface though, as that is important to the design.  Currently, this is used to figure out what transformation, if any, we need to do to get from a node in B to the corresponding node in A.  This would be a dequant (the only thing supported now), an `nn.Identity()` (not in yet, but easy to add), a quant (might be harder to add and not a p0), a cast to fp16, etc.

We can probably rename this enum to `NodeFirstInputType` to make it clear, since that is the only thing used today.  If in the future we need types of other inputs, or outputs, we can change accordingly.

vkuzo committed:
Update on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403095](https://our.internmc.facebook.com/intern/diff/D26403095)

[ghstack-poisoned]

vkuzo committed:
rebase on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403095](https://our.internmc.facebook.com/intern/diff/D26403095)

[ghstack-poisoned]

raghuramank100 commented on torch/quantization/ns/graph_passes.py:
```diff
@@ -299,7 +304,8 @@ def load_arg(a):
             # node_a's input (dequant, etc)
             dtype_cast_node = _insert_dtype_cast_after_node(
-                node_a, node_c, node_c.args[0], gm_a, gm_b, node_b.name + '_dtype_cast_')
+                node_a, node_c, node_c.args[0], gm_a, gm_b, graph_c,
```
Just to be sure: graph_c shares weights with graph a and b, so a user cannot mutate a and b and assume that c is unchanged,

vkuzo replied:
More specifically,
1. graph_c and graph_b both have gm_b as the parent module (a graph itself has no weights, it points to attributes of the parent module)
2. the shadowing logic adds references from gm_b to the necessary objects (modules, packed_params, attrs) of gm_a

vkuzo committed:
rebase on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403095](https://our.internmc.facebook.com/intern/diff/D26403095)

[ghstack-poisoned]

vkuzo committed:
fix asan on "NS for FX: add test for a simple sparsenn model"


Summary:

Adds a very simple toy sparsenn model, and enables
its inspection with the new NS APIs.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_compare_activations
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_sparsenn_shadow
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26403095](https://our.internmc.facebook.com/intern/diff/D26403095)

[ghstack-poisoned]
